making a few assertion to throw in StandardWellPrimaryVariables

GitPaean opened the pull request:
to get more outputting information and also avoid harsh termination due to assert.

They are technically numerical problem instead of programming mistakes.

GitPaean committed:
making a few assertion to throw in StandardWellPrimaryVariables

to get more outputting information and also avoid harsh termination due
to assert.

They are technically numerical problem instead of programming mistakes.

GitPaean commented:
jenkins build this please

GitPaean commented on opm/simulators/wells/StandardWellPrimaryVariables.cpp:
```diff
@@ -680,7 +697,10 @@ relaxationFactorFractionsProducer(const BVectorWell& dwells) const
             }
         }
-        assert(relaxation_factor >= 0.0 && relaxation_factor <= 1.0);
+        if (relaxation_factor <= 0.0 && relaxation_factor > 1.0) {
```
`<= 0.0` is intentional to see whether it actually causes problem.

bska replied:
> `<= 0.0` is intentional to see whether it actually causes problem.

Okay, but maybe you meant `f <= 0.0 || f > 1.0` instead?  As written, this condition is equivalent to `if (false)`.

GitPaean replied:
changed to f < 0.0 || f > 1.0 to reflect the original behavoir.

akva2 commented on opm/simulators/wells/StandardWellPrimaryVariables.cpp:
```diff
@@ -50,9 +53,15 @@ namespace {
+                                const Scalar dx,
+                                const std::string& well_name,
+                                const std::string& pv_name,
+                                Opm::DeferredLogger& deffered_logger)
```
one f

GitPaean replied:
thanks. will address it when jenkins is done.

bska replied:
> thanks. will address it when jenkins is done.

Cool.  Would you mind adding doxygen-style descriptions of the parameters too, please?  That might help future maintainers get a handle on the meainings of (e.g.) `pv_name` here.

bska reviewed (commented):
This looks good to me for the most part.  Just a couple of minor issue, including, I think, one misused logical operator.

GitPaean commented:
jenkins build this please

GitPaean commented:
jenkins build this please

GitPaean commented:
I am marking it as ready for review. And the comments are addressed.

totto82 reviewed (approved):
Thanks for the work.

atgeirr commented on opm/simulators/wells/StandardWellPrimaryVariables.cpp:
```diff
@@ -680,7 +705,10 @@ relaxationFactorFractionsProducer(const BVectorWell& dwells) const
         }
-        assert(relaxation_factor >= 0.0 && relaxation_factor <= 1.0);
+        if (relaxation_factor < 0.0 || relaxation_factor > 1.0) {
+            const std::string msg = fmt::format(" illegal relaxation_fractor {} is obtained for well {}", relaxation_factor, this->well_.name());
```
Typo: fractor -> factor. Also, avoid variable names in messages, prefer the term "relaxation factor" to the variable name "relaxation_factor".

GitPaean replied:
done addressing.

atgeirr reviewed (approved):
Only minor typo fix, otherwise fine. I did think a while about the second assert in relaxationFactorFraction(), line 89, and concluded that it is correct.

blattms commented:
Guys, I think you missed some asserts. e,g. [StandardWellPrimaryVariables.cpp#L89](https://github.com/OPM/opm-simulators/blob/17111bf71c1a11160660a27aa9e6ca22e8c7a9ae/opm/simulators/wells/StandardWellPrimaryVariables.cpp#L89)

blattms commented:
While I agree that a negative factor will break the simulation, I am a bit unsure whether we should really abort for something larger than 1. Getting something slightly larger than 1 might happen due to rounding errors.
Somebody should try what happens, if we just use 1 in this case, issue a warning and continue.

GitPaean commented:
> While I agree that a negative factor will break the simulation, I am a bit unsure whether we should really abort for something larger than 1. Getting something slightly larger than 1 might happen due to rounding errors.
Somebody should try what happens, if we just use 1 in this case, issue a warning and continue.

We can do 1 + EPSILON, no problem.

GitPaean commented:
I will change EPISLON to be `std::numeric_limits<Scalar>::epsilon();`.

GitPaean commented:
> > While I agree that a negative factor will break the simulation, I am a bit unsure whether we should really abort for something larger than 1. Getting something slightly larger than 1 might happen due to rounding errors.
> > Somebody should try what happens, if we just use 1 in this case, issue a warning and continue.
> 
> We can do 1 + EPSILON, no problem.

with  [StandardWellPrimaryVariables.cpp#L89](https://github.com/OPM/opm-simulators/blob/17111bf71c1a11160660a27aa9e6ca22e8c7a9ae/opm/simulators/wells/StandardWellPrimaryVariables.cpp#L89), we should not do 1 + EPSILON.

GitPaean commented:
There might be more that we can do in this parts.  Will keep [StandardWellPrimaryVariables.cpp#L89](https://github.com/OPM/opm-simulators/blob/17111bf71c1a11160660a27aa9e6ca22e8c7a9ae/opm/simulators/wells/StandardWellPrimaryVariables.cpp#L89), as it is for the moment. 

The original motive is that we could not reproduce a failure regarding `assert(old_value >= -EPSILON && old_value <= (1.0+EPSILON));`. 

Maybe with this PR, the people that reported the issue can have a little more output.  And also, maybe a cutting time step is also a good measure instead of terminating the simulation harshly. 

There might be more discussion in this topic with testing evolving.

GitPaean commented:
jenkins build this please

GitPaean commented:
I am merging this with the approvals and addressed comments.  As mentioned above, this is not a done story. There might be more similar adjustments with more testing.
